client/tinychain/ml/linalg.py

```diff
@@ -206,16 +216,56 @@ def cond(p: Tensor, l: Tensor, u: Tensor, i: UInt):
         l=identity(x.shape[0], F32).as_dense().copy(),
         u=x.copy(),
         i=UInt(0),
+        num_permutations=UInt(0),
         ))))
 
+@post_op
+def det(cxt, x: Tensor) -> F32:
+    """Computes the determinant of square `matrix`.
 
-def slogdet(x):
-    """Compute the sign and log of the absolute value of the determinant of one or more square matrices."""
+    Args:
+        `x`: a matrix with shape `[N, N]`
 
-    # TODO: implement slogdet, cf. https://www.tensorflow.org/api_docs/python/tf/linalg/slogdet
+    Returns:
+        The determinant for `x`
+    """
 
-    raise NotImplementedError(f"slogdet({x})")
+    cxt.plu = plu
+    plu_result = cxt.plu(x=x)
+    sign = Int(-1).pow(plu_result.num_permutations)
+    
+    return diagonal(plu_result.u).product()*sign
+
+@post_op
+def slogdet(cxt, x: Dense) -> Tuple:
+    """Compute the sign and log of the absolute value of the determinant of one or more square matrices.
+
+    Args:
+        `x`: a `Tensor` of square `matrix`es with shape `[N, M, M]`
+
+    Returns:
+        The `Tuple` of `Tensor`s `(sign, logdet)` where:
+        `sign`: a `Tensor` of signs of determinants `{-1, +1}` with shape `[N,]`
+        `logdet`: a `Tensor` of the natural log of the absolute values of determinants with shape `[N,]`
+    """
+
+    n = x.shape[0]
+    cxt.result = Dense.create([n, 2])
+    cxt.det = det
+
+    @closure
+    @get_op
+    def step(i: UInt):
+        d = F32(cxt.det(x = x[i]))
+        logdet = F32(d.abs().log())
+        sign = Int(If(d > 0, 1, -1))*1
+        return After(when=[
```

Comment:
Would it make more sense to return two tensors? This is what TensorFlow does: https://www.tensorflow.org/api_docs/python/tf/linalg/slogdet

Reply:
Sure. My `slogdet` return two `Tensor`s. `Slogdet` write results `(sing, logdet)` in one tensor result, but after `slogdet` return `Tuple(sign, logdet)` where `sing`: a `Tensor` and `logdet`: `Tensor`. I can change one result tensor on two independent tensors. What do you think about it?